README.md
[![Build Status](https://simona.ie3.e-technik.tu-dortmund.de/ci/buildStatus/icon?job=ie3-institute%2FDWDWeatherTools%2Fmaster)](https://simona.ie3.e-technik.tu-dortmund.de/ci/job/ie3-institute/job/DWDWeatherTools/job/master/)
[![Quality Gate Status](https://simona.ie3.e-technik.tu-dortmund.de/sonar/api/project_badges/measure?project=edu.ie3%3Adwdtools&metric=alert_status)](https://simona.ie3.e-technik.tu-dortmund.de/sonar/dashboard?id=edu.ie3%3Adwdtools)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/658808baceae464497d90562a61a09a4)](https://www.codacy.com/gh/ie3-institute/DWDWeatherTools?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ie3-institute/DWDWeatherTools&amp;utm_campaign=Badge_Grade)
[![codecov](https://codecov.io/gh/ie3-institute/DWDWeatherTools/branch/master/graph/badge.svg)](https://codecov.io/gh/ie3-institute/DWDWeatherTools)
[![License](https://img.shields.io/github/license/ie3-institute/powersystemdatamodel)](https://github.com/ie3-institute/powersystemdatamodel/blob/master/LICENSE)
[![OpenData](https://img.shields.io/badge/Open-Data-d3281a?style=flat)](https://okfn.org/opendata/)
[![OpenContent](https://img.shields.io/badge/Open-Content-637abd?style=flat)](https://okfn.org/opendata/)
[![OpenService](https://img.shields.io/badge/Open-Service-ff6600?style=flat)](https://okfn.org/opendata/)

# DWDWeatherTools
Java command line client to download and convert weather data of the
[ICON-EU](https://www.dwd.de/DE/leistungen/modellvorhersagedaten/modellvorhersagedaten.html)
model from the Deutscher Wetterdienst ([DWD](https://www.dwd.de/EN/Home/home_node.html)),
the federal meteorological service in Germany.

These tools are work in progress and although they are used in our institute to collect
data from the DWD they might contain bugs. If find one, please consider handing in an issue.

## Table of Contents

- [Background Information](#background-information)
- [Features](#features)
- [Processed Data](#processed-data)
- [Requirements](#requirements)
- [Usage](#usage)

### Background Information
The DWD provides a free weather forecast on their server using the ICON model.
The forecast consists of different forecast variables and parameters which may
have one or more height levels.

Every 3 hours the DWD starts a calculation for each parameter for the next
12-76 hours (depending on the parameter). Afterwards the model run results are uploaded,
which may take up to 5 hours after the start, but most times takes about 2-3 hours.
The results are structured into GRIB2 (.grib2) files, one per calculated hour and parameter,
that are compressed as .bz2-archives. The data normally is overwritten everytime new model run
data is uploaded. To keep data permanently e.g. for research purposes, one has to store them
locally and run the downloader recurring e.g by using a cron job every 3 hours.
The path to each result is built logically:

`https://opendata.dwd.de/weather/nwp/icon-eu/grib/<modelrun in HH-format>/<parameter_name>/icon-eu_europe_regular-lat-lon_[model/single]-level_<modelrun-date in YYYYMMDDHH-format>_<timestep in HHH-format>_<[height level in two-digit-format_]><parameter_name>.grib2.bz2`

### Features
-   Downloader to collect the temporarily available ICON weather data from the
	DWD OpenData servers
-   Converter to extract the weather data from the downloaded files and persist
	them into a PostgreSQL database

### Processed Data

#### Parameter
While in general the downloaded raw data might contain all values described in the model
[documentation](https://isabel.dwd.de/DWD/forschung/nwv/fepub/icon_database_main.pdf),
in our case we only need some specific variables and parameters, which are extracted
during the conversion process. If you require additional parameters you might consider
adapting the downloader and converter accordingly or hand in an issue.

At the moment, the following parameters are converted:

// todo parameters

#### Grid Information
The currently covered area by the DWD Icon model consists of the coordinates shown below.
This whole area is covered by the downloader as well as all parameters are downloaded for
the whole area.

|                   | ICON-EU nest                          |
|-------------------|---------------------------------------|
|geogr. coordinates | 23.5° W – 62.5° E / 29.5° N – 70.5° N |
|mesh size          | ≈ 6.5 km (R3B08) / 659156 triangles   |
|vertical levels	| 60 level                              |
|upper boundary	    | 22.5km                                |

![](docs/readthedocs/_static/img/icon-eu-nest.png)


In contrast, the converter area can be set in the code to the area of interest.
Hence, as long as the raw icon data is available, it can always be filtered and
converted into a specific area of interested. The current conversion area is set
to the following coordinates.

minLongitude = 4.29694 \
maxLongitude = 18.98635 \
minLatitude = 45.71457 \
maxLatitude = 57.65129

In the future it would make sense to provide the
coordinates via a config file instead of hard-coding them into the binary.
If you want to provide this future feel free to hand in an issue or a pull request.

![](docs/readthedocs/_static/img/converter_covered_area.jpg)


### Requirements
* Unix based OS
* [ecCodes](https://confluence.ecmwf.int//display/ECC/Releases) installed
* JRE 17 installed
* access to a postgresql database
* internet access
* write permission on download directory
* full privileges to the .jar file of the tools


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased Snapshot]

### Added
- Added ReadTheDocs support [#94](https://github.com/ie3-institute/DWDWeatherTools/issues/94)
- Added a changelog

### Changed

### Fixes


src/main/java/edu/ie3/tools/Main.java
/*
 * © 2019. TU Dortmund University,
 * Institute of Energy Systems, Energy Efficiency and Energy Economics,
 * Research group Distribution grid planning and operation
 */

package edu.ie3.tools;

import java.io.File;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;
import java.util.TimeZone;
import picocli.CommandLine;

@CommandLine.Command(
    name = "dwdtools",
    description =
        "Provides functions for the download, extraction and conversion of DWD GRIB2-Files",
    mixinStandardHelpOptions = true,
    version = "0.4",
    header =
        "\n"
            + "\n"
            + "██████╗ ██╗    ██╗██████╗     ████████╗ ██████╗  ██████╗ ██╗     ███████╗\n"
            + "██╔══██╗██║    ██║██╔══██╗    ╚══██╔══╝██╔═══██╗██╔═══██╗██║     ██╔════╝\n"
            + "██║  ██║██║ █╗ ██║██║  ██║       ██║   ██║   ██║██║   ██║██║     ███████╗\n"
            + "██║  ██║██║███╗██║██║  ██║       ██║   ██║   ██║██║   ██║██║     ╚════██║\n"
            + "██████╔╝╚███╔███╔╝██████╔╝       ██║   ╚██████╔╝╚██████╔╝███████╗███████║\n"
            + "╚═════╝  ╚══╝╚══╝ ╚═════╝        ╚═╝    ╚═════╝  ╚═════╝ ╚══════╝╚══════╝\n"
            + "                                                                         \n"
            + "\n\n",
    footer = {
      "\n\n© 2019 Technische Universität Dortmund",
      "Institut für Energiesysteme, Energieeffizienz und Energiewirtschaft",
      "Forschungsgruppe Verteilnetzplanung und -betrieb"
    })
public class Main {

  @CommandLine.Option(
      names = {"-d", "--dir", "--directory"},
      description =
          "The directory used to write downloads to or read files from. Default: downloads")
  public static String directory = "downloads";

  @CommandLine.Option(
      names = {"--timesteps"},
      description = "Number of timesteps ( = hours) to process in a modelrun. Default: 12")
  public static int timesteps = 12;

  @CommandLine.Option(
      names = {"-t", "--tolerance", "--fault_tolerance"},
      description = "Fault tolerance, e.g. missing coordinates. Default: 0.33")
  public static double faultTolerance = 0.33;

  @CommandLine.Option(
      names = {"convert"},
      description = "Converts grib2 files")
  public static boolean doConvert;

  @CommandLine.Option(
      names = {"-convert_from"},
      description = "Start datetime of conversion.")
  public static String convertFrom;

  @CommandLine.Option(
      names = {"-convert_until"},
      description = "End datetime of conversion.")
  public static String convertUntil;

  @CommandLine.Option(
      names = {"download"},
      description = "Downloads grib2 files")
  public static boolean doDownload;

  @CommandLine.Option(
      names = {"-i", "--interpolation_ratio"},
      description =
          "The ratio at which data is interpolated, at a value of 1, previous value will be overriden. Default: 0.67")
  public static double interpolationRatio = 0.67; // Overrides previous values if set to 1

  @CommandLine.Option(
      names = {"-del", "--delete"},
      description = "Delete downloaded files after conversion. Default: false")
  public static boolean deleteDownloadedFiles = false;

  @CommandLine.Option(
      names = {"-user", "--database_user"},
      description = "Database user")
  public static String databaseUser = "";

  @CommandLine.Option(
      names = {"-pwd", "--database_password"},
      description = "Database password")
  public static String databasePassword = "";

  @CommandLine.Option(
      names = {"-conn", "--connection_url"},
      description = "Database connection url. Default: jdbc:postgresql://127.0.0.1:5432/dwd")
  public static String connectionUrl = "jdbc:postgresql://127.0.0.1:5432/dwd";

  @CommandLine.Option(
      names = {"-schema"},
      description = "default database schema")
  public static String database_schema = "icon";

  @CommandLine.Option(
      names = {"-m", "--missing_value_string"},
      description = "Configure the NULL-Value-String to set and parse in textfile")
  public static String missingValue = "null";

  @CommandLine.Option(
      names = {"-eccodes", "--eccodes_location"},
      description = "The location of the eccodes commands")
  public static String eccodes = "/usr/local/bin/grib_get_data";

  @CommandLine.Option(
      names = {"-filestatus"},
      description = "Write file status changes into FileStatus.log")
  public static boolean filestatus = false;

  @CommandLine.Option(
      names = {"debug"},
      description = "Create debugging output")
  public static boolean debug = false;

  // Coordinate-Rectangle to convert
  public static final double MIN_LONGITUDE = 4.29694;
  public static final double MAX_LONGITUDE = 18.98635;
  public static final double MIN_LATITUDE = 45.71457;
  public static final double MAX_LATITUDE = 57.65129;

  public static void main(String[] args) {
    // ICON-EU data timezone is UTC time,
    // hence we need to enforce this timezone for the whole application
    TimeZone.setDefault(TimeZone.getTimeZone("UTC"));

    CommandLine commandLine = new CommandLine(new Main());
    commandLine.parseArgs(args);
    if (commandLine.isUsageHelpRequested() || !(doDownload || doConvert)) {
      commandLine.usage(System.out);
      return;
    } else if (commandLine.isVersionHelpRequested()) {
      commandLine.printVersionHelp(System.out);
      return;
    }
    directory = directory.replace("\"", "").replace("'", "");
    if (directory.endsWith(File.separator))
      directory = directory.substring(0, directory.length() - 1).trim();
    eccodes = eccodes.replace("\"", "").replace("'", "");
    if (eccodes.endsWith(File.separator))
      eccodes = eccodes.substring(0, eccodes.length() - 1).trim();
    if (!eccodes.endsWith("grib_get_data")) {
      if (!eccodes.isEmpty()) eccodes += File.separator;
      eccodes += "grib_get_data";
    }

    if (doConvert) {
      if (convertFrom != null && convertUntil != null) {
        DateTimeFormatter dateTimeFormatter =
            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
                .withZone(ZoneId.of("UTC"))
                .withLocale(Locale.GERMANY);
        ZonedDateTime from = ZonedDateTime.parse(convertFrom, dateTimeFormatter);
        ZonedDateTime until = ZonedDateTime.parse(convertUntil, dateTimeFormatter);
        new Converter(from, until).run();
      } else if (convertFrom != null || convertUntil != null) {
        throw new IllegalArgumentException(
            "Either convertFrom or convertTo is missing. We need both to convert data for a specific interval");
      } else {
        new Downloader().run();
      }
    }
  }

  public static Set<String> printProgramArguments() {
    return printProgramArguments(false);
  }

  public static Set<String> printProgramArguments(boolean verbose) {
    HashSet<String> args = new HashSet<>();
    args.add("directory = \"" + Main.directory + "\"");
    if (verbose) args.add("timesteps = " + timesteps);
    if (verbose) args.add("faultTolerance = " + faultTolerance);
    args.add("interpolationRatio = " + interpolationRatio);
    if (verbose) args.add("databaseUser = \"" + databaseUser + "\"");
    args.add("connectionUrl = \"" + connectionUrl + "\"");
    if (verbose) args.add("missingValue = \"" + missingValue + "\"");
    args.add("eccodes location= \"" + eccodes + "\"");
    return args;
  }
}
